### ISSUE TITLE
A Bug in SEMEstimator.py

### ISSUE THREAD

**Issue by jihyeseo** [pgmpy/pgmpy#1902]
The masks are being used in different ways in SEM.py and SEMEstimator.py

in [SEM.py](https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/models/SEM.py#L882C1-L886C10)
```
# Masks represent the parameters which need to be learnt while training.
self.B_mask = np.multiply(np.where(self.B_fixed_mask != 0, 0.0, 1.0), self.B)
self.zeta_mask = np.multiply(
    np.where(self.zeta_fixed_mask != 0, 0.0, 1.0), self.zeta
)
```

in [SEMEstimator.py](https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/estimators/SEMEstimator.py#L75)
```
B_masked = torch.mul(B, self.B_mask) + self.B_fixed_mask
B_inv = pinverse(self.B_eye - B_masked)
zeta_masked = torch.mul(zeta, self.zeta_mask) + self.zeta_fixed_mask
```
[and here](https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/estimators/SEMEstimator.py#L325C1-L327C1)
```
B = params["B"] * self.B_mask + self.B_fixed_mask
zeta = params["zeta"] * self.zeta_mask + self.zeta_fixed_mask
```

Using the test cases in `TestSEMEstimator`, 
`self.B_fixed_mask` have NaN's. Leading to failure in pinverse().

Do we expect the masking to act like SEM:
```
B = torch.mul(torch.where(self.B_fixed_mask != 0, 0.0, 1.0), params["B"])
zeta = torch.mul(torch.where(self.zeta_fixed_mask != 0, 0.0, 1.0), params["zeta"])
```
?

But still I could not make the [test cases](https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/tests/test_estimators/test_SEMEstimator.py#L101-L151) work, I was getting [negative determinant](https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/estimators/SEMEstimator.py#L343C20-L343C44) for `sigma_hat`.

**Comment by ankurankan** [comment]
@jihyeseo Thanks for looking into the issue. I think the masks are used correctly in SEM and SEMEstimator. In SEM.py, the idea is to create a mask for parameters that need to be learned, i.e., (parameters corresponding to the edges - edges that have fixed parameters). In the case of SEMEstimator, we are trying to recreate this parameter matrix, so we first throw out non-learnable parameters using the mask and then add back the fixed parameters. 

 I think the nan issue is coming from the method converting the SEMGraph model to SEMAlg model, specifically this line: https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/models/SEM.py#L592. Here, because of the `weight='weight'` argument, the parameters that are not fixed get nan assigned to them. We would essentially need to change this to first checking whether the "weight" attribute exists for an edge or not. If it exists we take the value, otherwise return 0. I imagine we might need to change this line similarly as well: https://github.com/pgmpy/pgmpy/blob/f4a0485668f437182d08abdf4b33d5be1b7c42d1/pgmpy/models/SEM.py#L596

**Comment by jihyeseo** [comment]
Thanks for the check and the explanation! 
I will try to fix this for a few days :-) Could you assign me?
